mender-auth/http_forwarder/http_forwarder_test.cpp

```diff
@@ -0,0 +1,536 @@
+			exp_req.value()->SetBodyWriter(writer);
+		},
+		[](http::ExpectedIncomingRequestPtr exp_req) { ASSERT_FALSE(exp_req); });
+
+	http::ClientConfig client_config;
+
+	hf::TestServer forwarder(server_config, client_config, loop);
+	auto err = forwarder.AsyncForward("http://127.0.0.1:0", "http://127.0.0.1:" TEST_PORT "/");
+	ASSERT_EQ(err, error::NoError);
+
+	http::Client client(client_config, loop);
+	auto req = make_shared<http::OutgoingRequest>();
+	req->SetMethod(http::Method::PUT);
+	req->SetAddress(http::JoinUrl(forwarder.GetUrl(), "/test-endpoint"));
+	req->SetHeader("Content-Length", to_string(BodyOfXes::TARGET_BODY_SIZE));
+	req->SetBodyGenerator([]() { return make_shared<BodyOfXes>(); });
+	err = client.AsyncCall(
+		req,
+		[&loop](http::ExpectedIncomingResponsePtr exp_resp) {
+			ASSERT_FALSE(exp_resp);
+			EXPECT_THAT(
+				exp_resp.error().String(), ::testing::HasSubstr("Connection reset by peer"));
+			loop.Stop();
+		},
+		[](http::ExpectedIncomingResponsePtr exp_resp) {
+			ASSERT_TRUE(false) << "Should not get here";
+		});
+	ASSERT_EQ(err, error::NoError);
+
+	loop.Run();
+}
+
+TEST(HttpForwarderTests, ClientTerminatesUpload) {
+	mtesting::TestEventLoop loop;
+
+	http::ServerConfig server_config;
+	http::Server server(server_config, loop);
+	vector<uint8_t> req_body;
+	bool hit_server_header_handler {false};
+	bool hit_server_body_handler {false};
+	server.AsyncServeUrl(
+		"http://127.0.0.1:" TEST_PORT,
+		[&req_body, &hit_server_header_handler](http::ExpectedIncomingRequestPtr exp_req) {
+			ASSERT_TRUE(exp_req) << exp_req.error().String();
+			auto writer = make_shared<io::ByteWriter>(req_body);
+			writer->SetUnlimited(true);
+			exp_req.value()->SetBodyWriter(writer);
+			hit_server_header_handler = true;
+		},
+		[&hit_server_body_handler, &loop](http::ExpectedIncomingRequestPtr exp_req) {
+			ASSERT_FALSE(exp_req);
+			hit_server_body_handler = true;
+			loop.Stop();
+		});
+
+	http::ClientConfig client_config;
+
+	hf::TestServer forwarder(server_config, client_config, loop);
+	auto err = forwarder.AsyncForward("http://127.0.0.1:0", "http://127.0.0.1:" TEST_PORT "/");
+	ASSERT_EQ(err, error::NoError);
+
+	class DisconnectAtEndReader : virtual public io::Reader {
+	public:
+		http::Client &client;
+		io::ReaderPtr reader;
+
+		DisconnectAtEndReader(http::Client &client, io::ReaderPtr reader) :
+			client {client},
+			reader {reader} {
+		}
+
+		io::ExpectedSize Read(
+			vector<uint8_t>::iterator start, vector<uint8_t>::iterator end) override {
+			auto result = reader->Read(start, end);
+			if (result && result.value() == 0) {
+				client.Cancel();
+			}
+			return result;
+		}
+	};
+
+	http::Client client(client_config, loop);
+	auto req = make_shared<http::OutgoingRequest>();
+	req->SetMethod(http::Method::PUT);
+	req->SetAddress(http::JoinUrl(forwarder.GetUrl(), "/test-endpoint"));
+	// Too big, same as termination.
+	req->SetHeader("Content-Length", to_string(BodyOfXes::TARGET_BODY_SIZE * 2));
```

Comment:
As this test is explicitly terminating, we don't need to oversize the header, do we?

Reply:
Sort of true, but if we cancel exactly at the body length, then there is a race condition. The request is a complete request then and the server may actually have time to reply before the cancel goes through. Unlikely, since it's going to write a whole MB, but I'd rather have it unambiguous.